downstairs/src/region.rs

```diff
@@ -1053,6 +1014,136 @@ impl Region {
+                        *r =
+                            extent.flush(flush_number, gen_number, job_id, &log)
+                    });
+                }
+            })
+        });
+
+        for result in results {
+            // If any extent flush failed, then return that as an error. Because
+            // the results were all collected above, each extent flush has
+            // completed at this point.
+            result?;
+        }
+        Ok(())
+    }
+
+    /// Flush extents using the `syncfs` syscall
+    ///
+    /// If using `omicron-build`, then we're running on illumos and the
+    /// region is backed by a ZFS dataset. Issue a `syncfs` syscall, which
+    /// will result in a `zfs_sync` to the entire region dataset.
+    ///
+    /// The codepath is roughly as follows
+    /// - The syscall calls `fdsync(.., FDSYNC_FS)`
+    /// - `fdsync` calls `VFS_SYNCFS` for the appropriate filesystem
+    /// - For ZFS, this calls `zfs_syncfs`
+    /// - `zfs_syncfs` calls `zfs_sync(vfsp, 0, cr)`
+    /// - `zfs_sync` calls `zil_commit(zfsvfs->z_log, 0)` if `zfsvfs->z_log` is
+    ///   present; it should always be present, because it's only set to `NULL`
+    ///   during `zfsvfs_teardown`.
+    ///
+    /// Quoth the `zil_commit` docstring:
+    ///
+    /// > If "foid" is zero, this means all "synchronous" and "asynchronous"
+    /// > txs, for all objects in the dataset, will be committed to stable
+    /// > storage prior to zil_commit() returning.
+    ///
+    /// The extents all live within the same dataset, so this is what we want!
+    #[cfg(feature = "omicron-build")]
+    fn flush_extents(
+        &mut self,
+        dirty_extents: &BTreeSet<ExtentId>,
+        flush_number: u64,
+        gen_number: u64,
+        job_id: JobId,
+    ) -> Result<(), CrucibleError> {
+        #[cfg(not(target_os = "illumos"))]
+        compile_error!("syncfs-based flush is only supported on illumos");
+        use std::os::fd::AsRawFd;
+
+        // Prepare to flush extents by writing metadata to files (if needed)
+        //
+        // Some extents may not _actually_ be dirty, so we'll skip them here
+        let mut flushed_extents = vec![];
+        let log = self.log.clone();
+        for eid in dirty_extents {
+            let extent = self.get_opened_extent_mut(*eid);
+            if extent.pre_flush(flush_number, gen_number, job_id, &log)? {
+                flushed_extents.push(*eid);
+            } else {
+                warn!(
+                    self.log,
+                    "extent {eid} was in dirty_extents but not actually dirty"
+                );
+            }
+        }
+
+        // Execute the syscall
+        let dir = self.dir.clone();
+
+        // Do the flush in a worker thread if possible, to avoid blocking the
+        // main tokio thread pool.
+        run_blocking(move || -> Result<(), CrucibleError> {
+            let file = File::open(&dir)?;
+            let rc = unsafe { syncfs(file.as_raw_fd()) };
```

Comment:
Think there'd be any benefit to keeping this fd around persistently instead of re-opening every flush?

I don't think I saw this at all on the flamegraphs so I think the answer is no